reasoner/resolution/answer/AnswerState.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright (C) 2020 Grakn Labs
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package grakn.core.reasoner.resolution.answer;
+
+import grakn.core.common.exception.GraknException;
+import grakn.core.concept.Concept;
+import grakn.core.concept.answer.ConceptMap;
+import graql.lang.pattern.variable.Reference;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import static grakn.common.util.Objects.className;
+import static grakn.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static grakn.core.common.exception.ErrorMessage.Pattern.INVALID_CASTING;
+
+public abstract class AnswerState {
+    private final ConceptMap conceptMap;
+
+    AnswerState(ConceptMap conceptMap) {
+        this.conceptMap = conceptMap;
+    }
+    protected ConceptMap conceptMap() {
+        return conceptMap;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        final AnswerState that = (AnswerState) o;
+        return conceptMap.equals(that.conceptMap);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(conceptMap);
+    }
+
+    public static class UpstreamVars {
+
+        public static class Partial extends AnswerState {
```

Comment:
last comment to make this go in: can we call this "Initial"? This would make the transition more clear i think: `Upstream.Initial` -> `Downstream.Mapped/Downstream.Unified` -> `Upstream.Derived`?